### ISSUE TITLE
analyze block-witholding ('selfish mining') attacks in light of Equihash performance

### ISSUE THREAD

**Issue by zookozcash** [zcash/zcash#764]
Litecoin has had 2.5-minute blocks all along (with 1 MB block size), and it doesn't seem to have suffered any problems because of it. Let's set Zcash's block interval to 2.5 minutes too.

I've asked the Zcash Technical Advisory Board privately if they think there would be a problem with this, and none of them raised a warning flag.

**Comment by zookozcash** [comment]
After all the extensive discussions we've had about this and #765, discussing risks involving long verification times compared to block times, I'm currently still in favor of this and #765. The risks seem uncertain, far-future, and fixable. The benefits are immediate and significant.

**Comment by defuse** [comment]
#320 is relevant when we're thinking about miner transaction processing times.

**Comment by defuse** [comment]
How will this interact with the GBP proof of work? Is GBP sufficiently fast for 2.5 minute block intervals, or is there a long, say, minute-long memory filling operation you have to do before you can start running the GBP algorithm?

**Comment by daira** [comment]
We decided to implement this and #765 on the testnet, and dare people to break it :-p

**Comment by ebfull** [comment]
Reminded: when we do this, we must modify the founders' reward code, because it is based on the assumption we have 10 minute blocks, and thus that the "first four years" end on block 210000 -- the first halving of the block reward.

**Comment by defuse** [comment]
There's an interaction between decreasing the block interval using Equishash for mining (#27). As you reduce the block interval or increase the difficulty of finding one Equihash solution, the expected number of nonces you need to iterate through before you win the block goes down. If that number is close to 1, then mining is unfair: the first person to solve the Equihash problem wins, so the miner with the fastest computer wins every time. If that number is high, then you win with a probability that's proportional to your relative computing power, which is what we want. So we need to tune our Equihash parameters and block interval to make sure that number is high enough.

**Comment by daira** [comment]
This also interacts with the block propagation time: if you've mined a block, then even if you submit it immediately (thus maximising your chance of winning *this* block reward), you have more time than other miners to start working on the next block. If it takes on average N Equihash solutions to win a block, and the miner with the biggest computer can do N solutions within the propagation time, that miner will always win.

With the combination of increasing the block size and the verification time for transactions (hence the propagation time for verified blocks), decreasing the block target time, and increasing the time to obtain a candidate PoW solution as input to the difficulty filter &mdash;all of which result from individually well-motivated decisions& mdash; we're at significant risk of enabling unfair mining we are not very careful.

**Comment by nathan-at-least** [comment]
#893 is about analyzing security of this parameter choice. This ticket is just about implementing it and does not block on analysis.

**Comment by nathan-at-least** [comment]
Remember: Our goal is "roughly identical in units/time" curve as per Bitcoin, so this change implies changing per-block reward amounts and halving interval counts.

**Comment by defuse** [comment]
**TL;DR: We should make sure our solver can be run at least *two* times within the target block interval. This will ensure no <50% miner can mine unfairly in the sense that they expect to solve 1 block before anyone else can even make an attempt.**

Here's some more analysis, in a simplified model. There's probably a way to see the conclusion without all this math, but oh well.

Let *s* be the number of Equihash solutions (i.e. candidate blocks to satisfy the difficulty criteria) the solver produces on average in a single run. Let *p* be the probability that an Equihash solution passes the difficulty check. Ignore block propagation times. Let *T<sub>slow</sub>* be the running time of the Equihash solver most miners are using (i.e. the best one available to the public). Let *T<sub>fast</sub> < T<sub>slow</sub>* be the running time a *secret* Equihash solver one mining entity has at their disposal (e.g. better optimized code or faster hardware). Suppose that the fast miner has *L* parallel instances of the *T<sub>fast</sub>* solver at their disposal.

The expected number of blocks the fast miner can solve *before anyone else on the network even finds their first Equihash solution* is T<sub>slow</sub>/T<sub>fast</sub> \* L \* s \* p. Mining is unfair in the "race to finish one computation" sense if T<sub>slow</sub>/T<sub>fast</sub> \* L \* s \* p >= 1. So fairness requires T<sub>slow</sub>/T<sub>fast</sub> \* L \* s \* p < 1.

The variable *p* is affected by two things: The total computational resources of the network (i.e. difficulty) and the speed of a single-run Equihash solver. If the Equihash solver is slower than the block interval, difficulty adjustment saturates p to 1 regardless of the network's computational resources, and the result is unfair. As the Equihash solver gets faster relative to the block interval, p shrinks without needing to grow the network, so smaller networks are more fair. But by how much?

Let G be the number of slow-mining machines on the network and B be the target block interval. Difficulty adjustment tries to set p such that 1/p = G \* B/T<sub>slow</sub> \* s + L \* B/T<sub>fast</sub> \* s, so let's assume that's the case. If you substitute the inverse of that thing's right hand side into T<sub>slow</sub>/T<sub>fast</sub> \* L \* s \* p < 1  and do a lot of algebra you get this condition:

> T<sub>slow</sub>/T<sub>fast</sub> < G/L \* B/T<sub>slow</sub> + B/T<sub>fast</sub>

Let's write r = T<sub>slow</sub>/T<sub>fast</sub> (the speedup of the fast implementation over the slow one), n = B/T<sub>slow</sub> (the number of equihash solutions a single "slow" machine can find within the target block interval), and x = L/G (the number of "fast" machines relative to the number of "slow" machines). We get:

> r < n/x + n/r

This is satisfied when x < 1 (which is safe to assume, since otherwise the fast miner controls 50% or more of the mining hardware) and r < n. Proof:

> r < n
> r < n + 1 (obvious)
> r < n + n/r (since n > r and both are positive)
> r < n/x + n/r (since x < 1 and n is positive)

So as long as we ensure that the speedup factor of the fastest *secret* Equihash implementation is no greater than the number of times the common one finishes within the block interval, then no "fast" miners' "expected number of blocks won before everyone else finishes one Equihash solution" will ever reach 1.

If we don't think anyone will in-secret beat the public's optimizations by more than a factor of *two*, then everything's fine as long as the public optimized code can run at least *twice* within the target block interval.

- *TODO:* Is `none of the "fast" miners' "expected number of blocks won before everyone else finishes one Equihash solution" will never reach 1` the right definition of fairness?

- *TODO:* It's likely that I majorly screwed up the math; check it.

**Comment by defuse** [comment]
Here's how to see it without all the math. Suppose there are 3N mining machines in the network. N fast and 2N slow. If the fast machines are twice as fast as the slow machines and all types of machines finish within the target interval, then difficulty will be set as though there are 4N of the slow machines in the network, and the fast machines make up 50% of the network's power. First, the fast ones spit out N candidates, which win the block with probability 1/4. If not, the slow ones spit out their first candidates (2N total) and so they win with probability 1/2, while at the same time the fast ones spit out their second candidates (N of them) which win with probability 1/4.

**Comment by defuse** [comment]
We still have to consider the effect on [selfish mining](https://bitcoinmagazine.com/articles/selfish-mining-a-25-attack-against-the-bitcoin-network-1383578440) attacks, though! 

Suppose the common Equihash solver running time is close to the block target time, and an adversary has a 2x faster implementation. There is a simple block withholding strategy that lets the adversary mine more efficiently than the rest of the network *with certainty*.

Call the probability of one of your fast rounds winning a block R, and suppose R is small enough that the probability of N fast rounds winning can be approximated N\*R. Since you can only get 2 sequential rounds in within a block interval, your proportion of the hashpower is X = ~2R.

Mine one fast round on your private chain. If you lose, keep mining and win the block with probability X. If you win (probability R), withhold the block for the duration of one of your fast rounds (minus a little bit) while mining on your private chain. Publish the block you won just before everyone else finishes their slow round (thus you win it with certainty, as they switch to mining that one\*). You win the block after it with probability ~3R = 1.5X (since you were able to make 3 fast round attempts in total).

With this strategy, every time you win a block, you're 1.5 times as likely to win the next one [except for (most of) the blocks you win using this strategy]. The strategy gets less effective the more times an Equihash solver can be run within a block interval.

\* - Does our current implementation do this? Or will it unoptimially finish running started solvers even after a new block arrives? Or *is* the optimal strategy to finish running your solver?

**Comment by ebfull** [comment]
The current implementation does not halt an equihash run when we receive a block, but it does halt in between equihash runs.

**Comment by str4d** [comment]
This behaviour will change to "halt solver run on new block immediately" once multithreading is added, and the solver can properly respond to interrupts.

**Comment by ebfull** [comment]
I'm going to keep this open because @defuse's comments here are incredibly important and we need to review all of this behavior once we finish our changes.

**Comment by defuse** [comment]
From the engineering meeting, one idea is to find out an upper bound on the equihash runtime's fraction of the block target time in order to be sufficiently-resistant to these attacks.

**Comment by nathan-at-least** [comment]
Let's figure out a "simplified ratio" between mean(?) equihash solution discovery time and block target mean time as a security goal, and then use that for our equihash-time:block-time ratio definition. Then later we can determine if any new improved attacks can violate this ratio's assumptions.

**Comment by defuse** [comment]
Daira suggests 1/4.

I'll keep analyzing it, trying to calculate the best an attacker can do for a given ratio.

**Comment by daira** [comment]
Overlaps with #893.

**Comment by defuse** [comment]
This interacts with #998: If an attacker with a faster Equihash implementation can game the difficulty adjustment algorithm to shorten the effective block interval, then that reduces the number of sequential equihash rounds everyone on the network can accomplish before the attacker wins, thus increasing the attacker's advantage, thus giving them more opportunity to game the difficulty adjustment, and so on...

For example, if the attacker's Equihash implementation always finishes faster than everyone else's, they might be able to abuse this death spiral until the difficulty is zero, and (if the difficulty adjustment algorithm is not secure) keep it at zero so that they win every block.

**Comment by defuse** [comment]
Another way it more benignly interacts with #998 is that miners with a sequentially-fast implementation can just mine something else while they wait for the difficulty to fluctuate down enough that they have a big advantage of winning.

**Comment by defuse** [comment]
I wrote a simulation of the attack in https://github.com/zcash/zcash/issues/764#issuecomment-224786125 generalized so that the attacker can be arbitrarily many blocks ahead, only broadcasting them when doing so is to their advantage. Whenever I needed to simplify something, I tried to do it in a way that gives the attacker a better advantage, so the results of the simulation should be an *upper bound* on the amount of advantage you can get by following that strategy.

**There are definitely bugs in my simulation that will invalidate these results, so it NEEDS review.** But right now it is telling me:

- If 100 equihashes fit into the block interval, the attacker controls 25% of the network, and the attacker's implementation is 2x faster, then their advantage is negligible (<0.05%).
- If 2 equihashes fit into the block interval, the attacker controls 25% of the network, and the attacker's implementation is 2x faster, then they win about 20% more blocks by using the strategy.
- If 4 equihashes fit into the block interval, the attacker controls 25% of the network, and the attacker's implementation is 2x faster, then they win about 10% more blocks by using the strategy.
- If 8 equihashes fit into the block interval, the attacker controls 25% of the network, and the attacker's implementation is 2x faster, then they win about 5% more blocks by using the strategy.
- If 8 equihashes fit into the block interval, the attacker controls 25% of the network, and the attacker's implementation is 5x faster, then they win about 3-4% more blocks by using the strategy.

(The last two bullet points seem odd and might indicate a bug in my simulation, but they aren't immediately contradictory because when the attacker has a faster implementation but still only controls 25% of the network, it must be the case that they have fewer implementations running in parallel, and their advantage could actually be lower).

I think if we're okay with giving a <=5% extra advantage to selfish mining, then we should be OK to aim for >=8 equihash runs per target block interval.

*Note that this simulation assumes instant block propagation, and so this advantage is **on top of** the advantage due to regular selfish mining, the kind that takes advantage of nonzero block propagation time.*

**Remember, all of the above is tentative until my simulation code gets lots of review.**

**Comment by defuse** [comment]
I forgot to link to the code, it's here: https://gist.github.com/defuse/d9918af2b161e64de2e5da9b6531764d

**Comment by defuse** [comment]
Added the engineering meeting agenda label because I want to get @daira and/or @str4d to find the bugs in my simulation, and we also need to make a decision about what to do about these results.

**Comment by defuse** [comment]
I found and fixed a bug that caused it to overestimate the attacker's advantage. I edited the original comment with the corrected numbers and conclusions.

**Comment by zookozcash** [comment]
Seems like what to do about these results is: we now have a constraint that the Equihash prover needs to complete in fewer than 1/4 of the target block time, i.e. 37 seconds when (as currently) target block time is 150 seconds.

**Comment by zookozcash** [comment]
Hey, this simulation acts as if every block takes exactly target block time. In reality the block times will be some kind of Poissonnish distribution with a floor set by the minimum solving time of the fastest miner. Does that difference between model and reality substantially change the bottom line result?

**Comment by daira** [comment]
I believe we should update the simulation to not assume instant propagation (see https://github.com/zcash/zcash/issues/764#issuecomment-207404652 ). It should be sufficient to just assume a constant, pessimistic, propagation time.

**Comment by defuse** [comment]
@zookozcash: It doesn't work like that. You can think of it like this: There's only two miners on the network. The Attacker Miner has a server farm of fast Equihash machines, and the Normal Miner has a server farm of slower Equihash machines. The miners give tell all of their machines what to do at the same time, so they're always in the same state, and they all spit out their solutions at the same time, but there's still the probability that those solutions pass the difficulty check. So the poissonish distribution should be simulated correctly.

I think all of the other assumptions I mentioned above only make the simulation overestimate an Attacker Miner's advantage using the strategy, since 1. there is no delay in controlling there own hardware, and 2. it gives them perfect knowledge (based on the time since the last block was broadcast) of how soon the Normal Miner's going to get the next batch of Equihash solutions.

**Comment by defuse** [comment]
It also assumes the difficulty is constant. This interacts with #998.

**Comment by zawy12** [comment]
Does the 37 come from 37=difficulty x (floor + luck x speed)? If it is, slow PCs will be unfairly penalized as difficulty rises from difficulty x floor being > 37. At some point my PC's difficulty x floor > 2.5 minutes.

**Comment by daira** [comment]
No, a solution run time less than 37 seconds (a quarter of 2.5 minutes) should be sufficient to reduce the unfairness to less than 5% independent of difficulty. (Also note that this is in an attack scenario; normally the unfairness will be much less than that.)

**Comment by nathan-at-least** [comment]
Note: I'm removing this from the engineering meeting agenda because I'm relying on @str4d, @daira, and @defuse to figure out:

a. how this affects PoW design (parameter selection and optimization trade-off choices), and
b. if this "bleeds out of the PoW scope" and impacts anything else.

If b. is true, put it back on the agenda.

**Comment by defuse** [comment]
The way I wrote the simulation doesn't make it easy to simulate block propagation time, but I think I see a way to overestimate the advantage an attacker can get from it. Let's assume the attacker's blocks propagate instantly, but everyone else's blocks take T seconds to spread throughout the entire network. Also, the attacker detects that someone else broadcasted a block instantly (but doesn't gain the ability to mine on it). If the rest of the network broadcasts a block, and the attacker broadcasts a block T - ɛ seconds later, then it overrides any normal block (even if it's propagated to 99% of the nodes), and everyone starts mining on the attacker's block. Clearly this is better for the attacker than reality: in reality the attacker would have less chance of winning the race the longer they wait, with the probability dropping to 0 by time T; here the probability stays 1 until time T.

With this extra ability, the the attacker doesn't have to broadcast a block to preempt the Normal Machines which are just about to finish. (A) They can wait and broadcast the block only if Normal Machines finish and actually win the block. They can also (B) wait up to T seconds to do so.

**I'm much less confident that my simulation code is correct now,** but here are the results (these should upper bound the best advantage when taking advantage of the regular selfish mining strategy *and* slow PoW selfish mining strategy):

With mining during propagation time ((A) and (B)), assuming the attacker can finish exactly 1 Equihash run within T seconds:

- Baseline: If 100 equihashes fit into the block interval, the attacker controls 25% of the network, and the attacker's implementation is 2x faster, then they win 34% more blocks.
- If 2 equihashes fit into the block interval, the attacker controls 25% of the network, and the attacker's implementation is 2x faster, then they win 44% more blocks (+10% over baseline).
- If 4 equihashes fit into the block interval, the attacker controls 25% of the network, and the attacker's implementation is 2x faster, then they win 41% more blocks (+7% over baseline).
- If 8 equihashes fit into the block interval, the attacker controls 25% of the network, and the attacker's implementation is 2x faster, then they win 37% more blocks (+3% over baseline).
- If 8 equihashes fit into the block interval, the attacker controls 25% of the network, and the attacker's implementation is 5x faster, then they win 47% more blocks (+13% over baseline).

Without mining during propagation time ((A) only):

- Baseline: If 100 equihashes fit into the block interval, the attacker controls 25% of the network, and the attacker's implementation is 2x faster, then they win 34% more blocks.
- If 2 equihashes fit into the block interval, the attacker controls 25% of the network, and the attacker's implementation is 2x faster, then they win 37% more blocks (+3% over baseline).
- If 4 equihashes fit into the block interval, the attacker controls 25% of the network, and the attacker's implementation is 2x faster, then they win 35% more blocks (+1% over baseline).
- If 8 equihashes fit into the block interval, the attacker controls 25% of the network, and the attacker's implementation is 2x faster, then they win 34% more blocks (+0% over baseline).
- If 8 equihashes fit into the block interval, the attacker controls 25% of the network, and the attacker's implementation is 5x faster, then they win 45% more blocks (+11% over baseline).

It's weird how the advantage jumps up a lot from 2x speed to 5x speed, that didn't happen before. Maybe my code is wrong.

**Comment by defuse** [comment]
Huh, in the (A)-only case:

- If 8 equihashes fit into the block interval, the attacker controls 25% of the network, and the attacker's implementation is **3x** faster, then they win 58% more blocks (+24% over baseline).

That doesn't make make sense.

**Comment by defuse** [comment]
Oh, I'm an idiot. I need to compare a 5x measurement to a 5x baseline, not a 2x baseline. Whoops.

**Comment by zookozcash** [comment]
I'm confused by the multiple different inputs that can vary here. Our [current implementation of Equihash](https://speed.z.cash/changes/?tre=10&rev=a8270035c0d3596a59af35bfdccfc7d807e2aed7&exe=1&env=1) takes ave 30s (min 15s, max 63s). Is that, Taylor, still so slow that it exacerbates block-withholding attacks? That's a worry. We're out of time for making changes though! We're (almost) in consensus-code-freeze and we're (already!) in security-auditing-freeze.

One cheap, safe change to mitigate this would be to back off the block times from 150s to 300s.

**Comment by daira** [comment]
I believe that the Equihash solution times could be reduced without changing the parameters from n=200, k=9, by either spending more optimisation effort on the elimination of partials, or colliding on more than 8 bits to reduce the number of false partials. (With the previous parameters, this step was not as important because there tended to be fewer false partials, and so not much effort was spent on optimising it.)

Whether we have time to do that before launch is another question. Backing off the target block time to 5m is indeed another possibility that could also help to reduce other potential latency problems.

**Comment by defuse** [comment]
Is the average equihash time 30s in the sense that 50% of the time the solver runs in <=30s? If so, I think we're okay to go ahead with those parameters and the 150s block time, at least with respect to the first kind of attack (https://github.com/zcash/zcash/issues/764#issuecomment-224786125, not taking advantage of propagation time).

My simulation assumes the equihash solver's running time is constant instead of normally distributed (or whatever), so that might invalidate my results in favor of thinking we're safe when we're not.

I'll revisit the simulation today.

**Comment by defuse** [comment]
(@jack explained the 30s thing to me in a pairing)

**Comment by defuse** [comment]
The next steps are:

- Un-revert my changes to the simulation that produced https://github.com/zcash/zcash/issues/764#issuecomment-235708640, and try setting the attacker speedup factor to 1. Re-run all of those experiments to find the "selfish mining advantage due to equihash", i.e. equihash's effect on old-fashioned selfish mining. Do that as a sanity check.

- Review that un-reverted code for mistakes and/or rewrite it in a clearer way and compare the answers to https://github.com/zcash/zcash/issues/764#issuecomment-235708640.

**Comment by zawy12** [comment]
The distribution of solves begins around 15 seconds, plus or minus a second, then you jump up about 10 or 15 seconds (not between 25 and 30, but almost exactly 25 or 30 depending on speed of machine) to the next bin that is focused on 25 or 30 seconds, plus or minus a second or two.  It jumps up again like this for several more bins, and spreads out at the higher bins.  As I recall the first bin is about 13%, then the next two will be about 26% each, then back down to about 13% and so on, you might guess 9% for the next bin.  There was not an increase in average speed for any of my 3 types of machines between z7 and z8, ranging from 40 to 70 seconds (pairs). I ran it 250 times on 3 machines.  The variance increased a lot from z7 to z8 as you probably know.  Just having a reliable percentage at 20 seconds and below may be all that's needed. Str4d had more to say about it here: https://forum.z.cash/t/z8-speeds-feeds/1152/9

**Comment by zawy12** [comment]
Since GPU cores operate at 1/2 to 1/3 the clock speed of CPU cores, I wonder if minimizing time-to-first solution is the right thing to do, if being biased towards CPUs is the goal.

**Comment by daira** [comment]
@zawy12: minimizing time-to-first solution is always the best strategy in this case, since no work toward subsequent solutions is lost.

**Comment by daira** [comment]
I suspect that #1239 addresses the remaining issue here, but note that miners who implement that optimization will have an advantage, which may be a concern if it isn't implemented in the 1.0 release.

**Comment by daira** [comment]
@defuse Did you do the things in https://github.com/zcash/zcash/issues/764#issuecomment-240573723 ? Please reassign to someone else if appropriate.

**Comment by nathan-at-least** [comment]
This is quite similar to #893. I'm moving it into 1.0.1, which means if any critical issues are revealed by it, we'll need an urgent network upgrade.

**Comment by daira** [comment]
This is no longer an issue now that Equihash solution runs are *much* faster than the expected block time.
